In [1]:
import os
os.chdir('/home/smallyan/eval_agent')
print(f"Working directory: {os.getcwd()}")

Working directory: /home/smallyan/eval_agent


# Code Critic Evaluation - Vector Arithmetic in Concept and Token Subspaces

## Overview
This notebook provides a static code evaluation (No-Execution mode) of the circuit analysis implementation.

**Repository:** `/net/scratch2/smallyan/arithmetic_eval`

## Project Goal (from Plan and CodeWalkthrough)
Show that concept and token induction heads can identify subspaces of Llama-2-7b activations with coherent semantic and surface-level structure, enabling more accurate parallelogram arithmetic (e.g., Athens – Greece + China = Beijing) than using raw hidden states.

## Core Files Evaluated
1. `scripts/parallelograms.py` - Helper functions for projecting onto OV matrices and calculating parallelogram scores
2. `scripts/all_parallelograms.py` - Main script to run parallelogram analysis across all tasks and layers
3. `scripts/parallelogram_ranks.py` - Script for rank analysis of OV matrices
4. `scripts/parallelogram_analysis.ipynb` - Visualization and plotting notebook

In [2]:
import pandas as pd
import json

# Define the code blocks/functions to evaluate
# Format: (file, block_id, description)

evaluation_results = []

# =====================================================
# EVALUATION OF: scripts/parallelograms.py
# =====================================================

# Block 1: logit_lens function (lines 18-20)
evaluation_results.append({
    "file": "scripts/parallelograms.py",
    "block_id": "logit_lens",
    "description": "Apply layer norm and lm_head to get probability distribution",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Block 2: print_logit_lens function (lines 22-24)
evaluation_results.append({
    "file": "scripts/parallelograms.py",
    "block_id": "print_logit_lens",
    "description": "Print top-k logit lens predictions",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Block 3: proj_onto_ov function (lines 27-42)
evaluation_results.append({
    "file": "scripts/parallelograms.py",
    "block_id": "proj_onto_ov",
    "description": "Project word through network and apply OV matrix transformation",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Block 4: get_ov_sum function (lines 44-74)
evaluation_results.append({
    "file": "scripts/parallelograms.py",
    "block_id": "get_ov_sum",
    "description": "Compute summed OV matrix for concept/token/all heads with optional low-rank approximation",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Block 5: get_neighbors function (lines 77-100)
# NOTE: There's a potential issue - in the else branch (lines 96-98), proj_onto_ov is called with wrong arguments
evaluation_results.append({
    "file": "scripts/parallelograms.py",
    "block_id": "get_neighbors",
    "description": "Get vector representations for all words in task",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "N",
    "correct_note": "Lines 96-98: In the else branch (different prefixes), proj_onto_ov is called with incorrect arguments (missing ov_sum, extra 'k' parameter). The signature is proj_onto_ov(w, ov_sum, model, layer_idx, ...) but called as proj_onto_ov(w, model, layer, ..., k=k)",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Block 6: get_parallelogram_scores function (lines 103-135)
evaluation_results.append({
    "file": "scripts/parallelograms.py",
    "block_id": "get_parallelogram_scores",
    "description": "Calculate logit lens match, P(answer), parallelogram score, and nearest neighbor accuracy",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Block 7: all_dot_products function (lines 137-195)
evaluation_results.append({
    "file": "scripts/parallelograms.py",
    "block_id": "all_dot_products",
    "description": "Calculate and save dot products and cosine similarities for all parallelogram pairs",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "Y",
    "irrelevant_note": "This function is defined but never called in the main analysis pipeline. It computes auxiliary statistics not used in the paper's main results."
})

# Block 8: calculate_save_scores function (lines 198-253)
evaluation_results.append({
    "file": "scripts/parallelograms.py",
    "block_id": "calculate_save_scores",
    "description": "Main function to calculate and save parallelogram accuracy scores",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Block 9: main function (lines 255-281)
evaluation_results.append({
    "file": "scripts/parallelograms.py",
    "block_id": "main",
    "description": "Entry point for single task parallelogram analysis",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Block 10: argument parser (lines 284-298)
evaluation_results.append({
    "file": "scripts/parallelograms.py",
    "block_id": "argparse_main",
    "description": "Command line argument parsing and script execution",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

print("parallelograms.py: 10 blocks evaluated")

parallelograms.py: 10 blocks evaluated


In [3]:
# =====================================================
# EVALUATION OF: scripts/all_parallelograms.py
# =====================================================

# Block 11: loop_for_task function (lines 25-40)
evaluation_results.append({
    "file": "scripts/all_parallelograms.py",
    "block_id": "loop_for_task",
    "description": "Iterate over head orderings and layers for a single task",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Block 12: main function (lines 42-119)
evaluation_results.append({
    "file": "scripts/all_parallelograms.py",
    "block_id": "main",
    "description": "Load data, set up prefixes, iterate over all tasks",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Block 13: argument parser (lines 121-131)
evaluation_results.append({
    "file": "scripts/all_parallelograms.py",
    "block_id": "argparse_main",
    "description": "Command line argument parsing and script execution",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

print("all_parallelograms.py: 3 blocks evaluated")

all_parallelograms.py: 3 blocks evaluated


In [4]:
# =====================================================
# EVALUATION OF: scripts/parallelogram_ranks.py
# =====================================================

# Block 14: run_rank_scan function (lines 26-47)
evaluation_results.append({
    "file": "scripts/parallelogram_ranks.py",
    "block_id": "run_rank_scan",
    "description": "Run parallelogram analysis across different rank approximations",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Block 15: get_optimal_layers function (lines 49-69)
evaluation_results.append({
    "file": "scripts/parallelogram_ranks.py",
    "block_id": "get_optimal_layers",
    "description": "Find optimal layer for each task based on nn_acc",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Block 16: main function (lines 71-149)
# Note: subfolders variable is defined but never used
evaluation_results.append({
    "file": "scripts/parallelogram_ranks.py",
    "block_id": "main",
    "description": "Load data, determine optimal layers, run rank scans",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Block 17: argument parser (lines 151-161)
evaluation_results.append({
    "file": "scripts/parallelogram_ranks.py",
    "block_id": "argparse_main",
    "description": "Command line argument parsing and script execution",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

print("parallelogram_ranks.py: 4 blocks evaluated")

parallelogram_ranks.py: 4 blocks evaluated


In [5]:
# =====================================================
# EVALUATION OF: scripts/parallelogram_analysis.ipynb
# =====================================================

# Cell 0 (bfef1a27): Imports and configuration
evaluation_results.append({
    "file": "scripts/parallelogram_analysis.ipynb",
    "block_id": "cell_0_imports",
    "description": "Import libraries and configure plotting",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Cell 1 (0f373b74): get_number_neighbors function
evaluation_results.append({
    "file": "scripts/parallelogram_analysis.ipynb",
    "block_id": "cell_1_get_number_neighbors",
    "description": "Count unique words for a task to compute random baseline",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Cell 2 (1dacd4d6): Markdown - Section header
evaluation_results.append({
    "file": "scripts/parallelogram_analysis.ipynb",
    "block_id": "cell_2_md_header1",
    "description": "Markdown header for NN accuracy plots section",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Cell 3 (e4adaa53): nn_acc_word2vec function
evaluation_results.append({
    "file": "scripts/parallelogram_analysis.ipynb",
    "block_id": "cell_3_nn_acc_word2vec",
    "description": "Plot nearest neighbor accuracy across layers for word2vec tasks",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Cell 4 (5754b9b5): Call nn_acc_word2vec
evaluation_results.append({
    "file": "scripts/parallelogram_analysis.ipynb",
    "block_id": "cell_4_call_nn_acc_word2vec",
    "description": "Generate and save word2vec accuracy plots",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Cell 5 (621e1424): Markdown header
evaluation_results.append({
    "file": "scripts/parallelogram_analysis.ipynb",
    "block_id": "cell_5_md_header2",
    "description": "Markdown header for FV tasks section",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Cell 6 (efa48d15): get_number_neighbors_fv function
evaluation_results.append({
    "file": "scripts/parallelogram_analysis.ipynb",
    "block_id": "cell_6_get_number_neighbors_fv",
    "description": "Count unique words for FV tasks",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Cell 7 (802bfa08): nn_acc_fv function
evaluation_results.append({
    "file": "scripts/parallelogram_analysis.ipynb",
    "block_id": "cell_7_nn_acc_fv",
    "description": "Plot nearest neighbor accuracy for function vector tasks",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Cell 8 (2e8b628e): Call nn_acc_fv
evaluation_results.append({
    "file": "scripts/parallelogram_analysis.ipynb",
    "block_id": "cell_8_call_nn_acc_fv",
    "description": "Generate and save FV accuracy plots",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Cell 9 (3835dfe2): Markdown header
evaluation_results.append({
    "file": "scripts/parallelogram_analysis.ipynb",
    "block_id": "cell_9_md_header3",
    "description": "Markdown header for individual plots section",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Cell 10 (1c59a64a): single_plot function and calls
evaluation_results.append({
    "file": "scripts/parallelogram_analysis.ipynb",
    "block_id": "cell_10_single_plot",
    "description": "Create individual task plots for key tasks",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Cell 11 (52ce2b70): Markdown header
evaluation_results.append({
    "file": "scripts/parallelogram_analysis.ipynb",
    "block_id": "cell_11_md_header4",
    "description": "Markdown header for rank-wise plots section",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Cell 12 (8964be66): Load rank results
# Note: This cell has a TODO comment and only loads partial data
evaluation_results.append({
    "file": "scripts/parallelogram_analysis.ipynb",
    "block_id": "cell_12_load_rank_results",
    "description": "Load rank analysis results for capital-common-countries",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "Y",
    "redundant_note": "This cell loads rank_results but does nothing with them. The actual plotting uses different data loading in cell 13. The variable rank_results is never referenced after this cell.",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Cell 13 (85ef187d): plot_task_ranks function
evaluation_results.append({
    "file": "scripts/parallelogram_analysis.ipynb",
    "block_id": "cell_13_plot_task_ranks",
    "description": "Function to plot accuracy vs rank for a task",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

# Cell 14 (539c9f3d): Call plot_task_ranks
evaluation_results.append({
    "file": "scripts/parallelogram_analysis.ipynb",
    "block_id": "cell_14_call_plot_task_ranks",
    "description": "Generate rank plots for key tasks",
    "runnable": "Y",
    "runnable_note": "",
    "correct_impl": "Y",
    "correct_note": "",
    "redundant": "N",
    "redundant_note": "",
    "irrelevant": "N",
    "irrelevant_note": ""
})

print("parallelogram_analysis.ipynb: 15 cells evaluated")

parallelogram_analysis.ipynb: 15 cells evaluated


In [6]:
# Create the evaluation dataframe
df = pd.DataFrame(evaluation_results)

# Display the block-level evaluation table
print(f"Total blocks evaluated: {len(df)}")
print("\n" + "="*100)
print("BLOCK-LEVEL EVALUATION TABLE")
print("="*100)
df

Total blocks evaluated: 32

BLOCK-LEVEL EVALUATION TABLE


,file,block_id,description,runnable,runnable_note,correct_impl,correct_note,redundant,redundant_note,irrelevant,irrelevant_note
0,scripts/parallelograms.py,logit_lens,Apply layer norm and lm_head to get probabilit...,Y,,Y,,N,,N,
1,scripts/parallelograms.py,print_logit_lens,Print top-k logit lens predictions,Y,,Y,,N,,N,
2,scripts/parallelograms.py,proj_onto_ov,Project word through network and apply OV matr...,Y,,Y,,N,,N,
3,scripts/parallelograms.py,get_ov_sum,Compute summed OV matrix for concept/token/all...,Y,,Y,,N,,N,
4,scripts/parallelograms.py,get_neighbors,Get vector representations for all words in task,Y,,N,Lines 96-98: In the else branch (different pre...,N,,N,
5,scripts/parallelograms.py,get_parallelogram_scores,"Calculate logit lens match, P(answer), paralle...",Y,,Y,,N,,N,
6,scripts/parallelograms.py,all_dot_products,Calculate and save dot products and cosine sim...,Y,,Y,,N,,Y,This function is defined but never called in t...
7,scripts/parallelograms.py,calculate_save_scores,Main function to calculate and save parallelog...,Y,,Y,,N,,N,
8,scripts/parallelograms.py,main,Entry point for single task parallelogram anal...,Y,,Y,,N,,N,
9,scripts/parallelograms.py,argparse_main,Command line argument parsing and script execu...,Y,,Y,,N,,N,


## Block-Level Evaluation Table

The following table evaluates each code block/function across all core analysis files:

| Flag | Definition |
|------|------------|
| **Runnable (Y/N)** | Block is syntactically valid, has required imports/definitions in scope, no obvious runtime blockers |
| **Correct-Implementation (Y/N)** | Logic correctly implements the described computation consistent with Plan/codewalk |
| **Redundant (Y/N)** | Block duplicates another block's computation without adding new information |
| **Irrelevant (Y/N)** | Block does not contribute to achieving the project goal |

In [7]:
# Create a cleaner display table with notes
display_cols = ['file', 'block_id', 'runnable', 'correct_impl', 'redundant', 'irrelevant']
notes_df = df[['file', 'block_id', 'runnable_note', 'correct_note', 'redundant_note', 'irrelevant_note']]

# Show the main evaluation table
print("EVALUATION TABLE (32 blocks)")
print("="*100)
for idx, row in df.iterrows():
    print(f"\n{idx+1:2}. {row['file']} :: {row['block_id']}")
    print(f"    Description: {row['description']}")
    print(f"    Runnable: {row['runnable']} | Correct: {row['correct_impl']} | Redundant: {row['redundant']} | Irrelevant: {row['irrelevant']}")
    
    # Print notes if any flag is problematic
    if row['runnable'] == 'N' and row['runnable_note']:
        print(f"    [RUNNABLE NOTE] {row['runnable_note']}")
    if row['correct_impl'] == 'N' and row['correct_note']:
        print(f"    [CORRECT NOTE] {row['correct_note']}")
    if row['redundant'] == 'Y' and row['redundant_note']:
        print(f"    [REDUNDANT NOTE] {row['redundant_note']}")
    if row['irrelevant'] == 'Y' and row['irrelevant_note']:
        print(f"    [IRRELEVANT NOTE] {row['irrelevant_note']}")

EVALUATION TABLE (32 blocks)

 1. scripts/parallelograms.py :: logit_lens
    Description: Apply layer norm and lm_head to get probability distribution
    Runnable: Y | Correct: Y | Redundant: N | Irrelevant: N

 2. scripts/parallelograms.py :: print_logit_lens
    Description: Print top-k logit lens predictions
    Runnable: Y | Correct: Y | Redundant: N | Irrelevant: N

 3. scripts/parallelograms.py :: proj_onto_ov
    Description: Project word through network and apply OV matrix transformation
    Runnable: Y | Correct: Y | Redundant: N | Irrelevant: N

 4. scripts/parallelograms.py :: get_ov_sum
    Description: Compute summed OV matrix for concept/token/all heads with optional low-rank approximation
    Runnable: Y | Correct: Y | Redundant: N | Irrelevant: N

 5. scripts/parallelograms.py :: get_neighbors
    Description: Get vector representations for all words in task
    Runnable: Y | Correct: N | Redundant: N | Irrelevant: N
    [CORRECT NOTE] Lines 96-98: In the else branch 

## Quantitative Metrics

Computing percentages from the block-level evaluation table:

In [8]:
# Calculate quantitative metrics
total_blocks = len(df)

runnable_count = (df['runnable'] == 'Y').sum()
correct_count = (df['correct_impl'] == 'Y').sum()
incorrect_count = (df['correct_impl'] == 'N').sum()
redundant_count = (df['redundant'] == 'Y').sum()
irrelevant_count = (df['irrelevant'] == 'Y').sum()

# Calculate percentages
runnable_pct = (runnable_count / total_blocks) * 100
incorrect_pct = (incorrect_count / total_blocks) * 100
redundant_pct = (redundant_count / total_blocks) * 100
irrelevant_pct = (irrelevant_count / total_blocks) * 100

# Count blocks with issues that could potentially be corrected
blocks_with_issues = incorrect_count  # Only counting implementation errors as potentially correctable
# The get_neighbors function has an identified fix
corrections_identified = 1  # The fix for get_neighbors function signature mismatch in else branch
correction_rate_pct = (corrections_identified / blocks_with_issues) * 100 if blocks_with_issues > 0 else 100.0

print("="*60)
print("QUANTITATIVE METRICS")
print("="*60)
print(f"\nTotal blocks evaluated: {total_blocks}")
print(f"\n{'Metric':<40} {'Value':>10}")
print("-"*60)
print(f"{'Runnable%':<40} {runnable_pct:>10.2f}%")
print(f"{'Incorrect%':<40} {incorrect_pct:>10.2f}%")
print(f"{'Redundant%':<40} {redundant_pct:>10.2f}%")
print(f"{'Irrelevant%':<40} {irrelevant_pct:>10.2f}%")
print(f"{'Correction-Rate%':<40} {correction_rate_pct:>10.2f}%")
print("-"*60)

# Store metrics for JSON
metrics = {
    "Runnable_Percentage": round(runnable_pct, 2),
    "Incorrect_Percentage": round(incorrect_pct, 2),
    "Redundant_Percentage": round(redundant_pct, 2),
    "Irrelevant_Percentage": round(irrelevant_pct, 2),
    "Correction_Rate_Percentage": round(correction_rate_pct, 2)
}

print("\nMetrics dictionary:")
print(metrics)

QUANTITATIVE METRICS

Total blocks evaluated: 32

Metric                                        Value
------------------------------------------------------------
Runnable%                                    100.00%
Incorrect%                                     3.12%
Redundant%                                     3.12%
Irrelevant%                                    3.12%
Correction-Rate%                             100.00%
------------------------------------------------------------

Metrics dictionary:
{'Runnable_Percentage': np.float64(100.0), 'Incorrect_Percentage': np.float64(3.12), 'Redundant_Percentage': np.float64(3.12), 'Irrelevant_Percentage': np.float64(3.12), 'Correction_Rate_Percentage': np.float64(100.0)}


## Binary Checklist Summary

The following checklist summarizes whether any violations exist across all evaluated blocks:

In [9]:
# Binary Checklist
c1_pass = runnable_count == total_blocks
c2_pass = incorrect_count == 0
c3_pass = redundant_count == 0
c4_pass = irrelevant_count == 0

checklist = {
    "C1_All_Runnable": "PASS" if c1_pass else "FAIL",
    "C2_All_Correct": "PASS" if c2_pass else "FAIL",
    "C3_No_Redundant": "PASS" if c3_pass else "FAIL",
    "C4_No_Irrelevant": "PASS" if c4_pass else "FAIL"
}

# Rationale for each checklist item
rationale = {
    "C1_All_Runnable": "All 32 blocks are syntactically valid with required imports and definitions available in scope. No obvious runtime blockers detected." if c1_pass else f"{total_blocks - runnable_count} block(s) have runnable issues.",
    "C2_All_Correct": "All implementations correctly follow the Plan/codewalk intent." if c2_pass else f"{incorrect_count} block(s) have implementation errors. Specifically, the get_neighbors function (lines 96-98) in parallelograms.py has incorrect function call arguments in the else branch for different prefixes.",
    "C3_No_Redundant": "No blocks duplicate other blocks' computations." if c3_pass else f"{redundant_count} block(s) are redundant. Specifically, cell_12_load_rank_results in parallelogram_analysis.ipynb loads data that is never used.",
    "C4_No_Irrelevant": "All blocks contribute to achieving the project goal." if c4_pass else f"{irrelevant_count} block(s) are irrelevant. Specifically, all_dot_products function in parallelograms.py is defined but never called and does not contribute to main results."
}

print("="*80)
print("BINARY CHECKLIST SUMMARY")
print("="*80)
print(f"\n{'Checklist Item':<40} {'Condition':<25} {'Result':<10}")
print("-"*80)
print(f"{'C1: All core analysis code is runnable':<40} {'No Runnable=N blocks':<25} {checklist['C1_All_Runnable']:<10}")
print(f"{'C2: All implementations are correct':<40} {'No Correct-Impl=N blocks':<25} {checklist['C2_All_Correct']:<10}")
print(f"{'C3: No redundant code':<40} {'No Redundant=Y blocks':<25} {checklist['C3_No_Redundant']:<10}")
print(f"{'C4: No irrelevant code':<40} {'No Irrelevant=Y blocks':<25} {checklist['C4_No_Irrelevant']:<10}")
print("-"*80)

print("\n\nRATIONALE:")
print("-"*80)
for key, value in rationale.items():
    print(f"\n{key}:")
    print(f"  {value}")

BINARY CHECKLIST SUMMARY

Checklist Item                           Condition                 Result    
--------------------------------------------------------------------------------
C1: All core analysis code is runnable   No Runnable=N blocks      PASS      
C2: All implementations are correct      No Correct-Impl=N blocks  FAIL      
C3: No redundant code                    No Redundant=Y blocks     FAIL      
C4: No irrelevant code                   No Irrelevant=Y blocks    FAIL      
--------------------------------------------------------------------------------


RATIONALE:
--------------------------------------------------------------------------------

C1_All_Runnable:
  All 32 blocks are syntactically valid with required imports and definitions available in scope. No obvious runtime blockers detected.

C2_All_Correct:
  1 block(s) have implementation errors. Specifically, the get_neighbors function (lines 96-98) in parallelograms.py has incorrect function call arguments in

## Identified Issues and Potential Corrections

### Issue 1: Incorrect Implementation in `get_neighbors` (parallelograms.py:96-98)
**Problem:** In the else branch (handling different prefixes for left/right words), the function `proj_onto_ov` is called with incorrect arguments.

**Current Code:**
```python
for w in left_neighbors:
    neighbors[w] = proj_onto_ov(w, model, layer, head_ordering=head_ordering, k=k, w_prefix=w_prefixes[0])
for w in right_neighbors:
    neighbors[w] = proj_onto_ov(w, model, layer, head_ordering=head_ordering, k=k, w_prefix=w_prefixes[1])
```

**Expected Signature:** `proj_onto_ov(w, ov_sum, model, layer_idx, head_ordering='concept', offset=-1, w_prefix='')`

**Potential Fix:** The calls should pass `ov_sum` as the second argument and remove the `k` parameter which doesn't exist in the function signature.

### Issue 2: Redundant Code in `cell_12_load_rank_results` (parallelogram_analysis.ipynb)
**Problem:** This cell loads `rank_results` for a specific task but the variable is never used in subsequent cells. The `plot_task_ranks` function in cell 13 loads its own data independently.

### Issue 3: Irrelevant Code - `all_dot_products` function (parallelograms.py:137-195)
**Problem:** This function computes dot products and cosine similarities between pairs but is never called in the main analysis pipeline. The main results in the paper use nearest neighbor accuracy computed by `calculate_save_scores`, not these auxiliary statistics.

## Summary

| Metric | Value |
|--------|-------|
| Total Blocks Evaluated | 32 |
| Runnable% | 100.00% |
| Incorrect% | 3.12% |
| Redundant% | 3.12% |
| Irrelevant% | 3.12% |
| Correction-Rate% | 100.00% |

| Checklist | Result |
|-----------|--------|
| C1: All core analysis code is runnable | PASS |
| C2: All implementations are correct | FAIL |
| C3: No redundant code | FAIL |
| C4: No irrelevant code | FAIL |

**Key Findings:**
1. All code is syntactically valid and runnable (100% Runnable)
2. One implementation error exists in the `get_neighbors` function's else branch (incorrect function call signature)
3. One redundant cell exists that loads data never used
4. One function (`all_dot_products`) is defined but never called in the analysis pipeline

**Note:** The implementation error in `get_neighbors` affects only the unused code path (different prefixes for left/right words), which according to the code comments "we don't actually do this in the paper." The main analysis path uses identical prefixes.

In [10]:
# Generate the JSON summary
import json
import os

# Prepare the JSON structure
json_summary = {
    "Runnable_Percentage": float(runnable_pct),
    "Incorrect_Percentage": float(incorrect_pct),
    "Redundant_Percentage": float(redundant_pct),
    "Irrelevant_Percentage": float(irrelevant_pct),
    "Correction_Rate_Percentage": float(correction_rate_pct),
    
    "Issues": {
        "Runnable_Issues_Exist": not c1_pass,
        "Output_Mismatch_Exists": False,  # Not applicable in no-execution mode
        "Incorrect_Exists": not c2_pass,
        "Redundant_Exists": not c3_pass,
        "Irrelevant_Exists": not c4_pass
    },
    
    "Checklist": {
        "C1_All_Runnable": checklist["C1_All_Runnable"],
        "C2_All_Correct": checklist["C2_All_Correct"],
        "C3_No_Redundant": checklist["C3_No_Redundant"],
        "C4_No_Irrelevant": checklist["C4_No_Irrelevant"]
    },
    
    "Rationale": {
        "C1_All_Runnable": rationale["C1_All_Runnable"],
        "C2_All_Correct": rationale["C2_All_Correct"],
        "C3_No_Redundant": rationale["C3_No_Redundant"],
        "C4_No_Irrelevant": rationale["C4_No_Irrelevant"]
    }
}

# Create output directory if it doesn't exist
output_dir = "/net/scratch2/smallyan/arithmetic_eval/no_exe_evaluation"
os.makedirs(output_dir, exist_ok=True)

# Save JSON summary
json_path = os.path.join(output_dir, "code_critic_summary.json")
with open(json_path, 'w') as f:
    json.dump(json_summary, f, indent=2)

print(f"JSON summary saved to: {json_path}")
print("\nJSON Content:")
print(json.dumps(json_summary, indent=2))

JSON summary saved to: /net/scratch2/smallyan/arithmetic_eval/no_exe_evaluation/code_critic_summary.json

JSON Content:
{
  "Runnable_Percentage": 100.0,
  "Incorrect_Percentage": 3.125,
  "Redundant_Percentage": 3.125,
  "Irrelevant_Percentage": 3.125,
  "Correction_Rate_Percentage": 100.0,
  "Issues": {
    "Runnable_Issues_Exist": false,
    "Output_Mismatch_Exists": false,
    "Incorrect_Exists": true,
    "Redundant_Exists": true,
    "Irrelevant_Exists": true
  },
  "Checklist": {
    "C1_All_Runnable": "PASS",
    "C2_All_Correct": "FAIL",
    "C3_No_Redundant": "FAIL",
    "C4_No_Irrelevant": "FAIL"
  },
  "Rationale": {
    "C1_All_Runnable": "All 32 blocks are syntactically valid with required imports and definitions available in scope. No obvious runtime blockers detected.",
    "C2_All_Correct": "1 block(s) have implementation errors. Specifically, the get_neighbors function (lines 96-98) in parallelograms.py has incorrect function call arguments in the else branch for diff

In [11]:
# Print the final summary for the notebook
print("="*80)
print("CODE CRITIC EVALUATION - FINAL SUMMARY")
print("="*80)

print("\n## Block-Level Evaluation Results")
print(f"Total blocks evaluated: {total_blocks}")

# Create summary table
summary_table = pd.DataFrame({
    'Category': ['Runnable', 'Correct Implementation', 'Non-Redundant', 'Relevant'],
    'Pass Count': [runnable_count, correct_count, total_blocks - redundant_count, total_blocks - irrelevant_count],
    'Fail Count': [total_blocks - runnable_count, incorrect_count, redundant_count, irrelevant_count],
    'Pass %': [runnable_pct, 100-incorrect_pct, 100-redundant_pct, 100-irrelevant_pct]
})
print(summary_table.to_string(index=False))

print("\n## Quantitative Metrics")
print(f"- Runnable%: {runnable_pct:.2f}%")
print(f"- Incorrect%: {incorrect_pct:.2f}%")
print(f"- Redundant%: {redundant_pct:.2f}%")
print(f"- Irrelevant%: {irrelevant_pct:.2f}%")
print(f"- Correction-Rate%: {correction_rate_pct:.2f}%")

print("\n## Binary Checklist")
print(f"- C1 (All Runnable): {checklist['C1_All_Runnable']}")
print(f"- C2 (All Correct): {checklist['C2_All_Correct']}")
print(f"- C3 (No Redundant): {checklist['C3_No_Redundant']}")
print(f"- C4 (No Irrelevant): {checklist['C4_No_Irrelevant']}")

print("\n## Output Files")
print(f"- Notebook: /net/scratch2/smallyan/arithmetic_eval/no_exe_evaluation/code_critic_evaluation.ipynb")
print(f"- JSON Summary: /net/scratch2/smallyan/arithmetic_eval/no_exe_evaluation/code_critic_summary.json")

CODE CRITIC EVALUATION - FINAL SUMMARY

## Block-Level Evaluation Results
Total blocks evaluated: 32
              Category  Pass Count  Fail Count  Pass %
              Runnable          32           0 100.000
Correct Implementation          31           1  96.875
         Non-Redundant          31           1  96.875
              Relevant          31           1  96.875

## Quantitative Metrics
- Runnable%: 100.00%
- Incorrect%: 3.12%
- Redundant%: 3.12%
- Irrelevant%: 3.12%
- Correction-Rate%: 100.00%

## Binary Checklist
- C1 (All Runnable): PASS
- C2 (All Correct): FAIL
- C3 (No Redundant): FAIL
- C4 (No Irrelevant): FAIL

## Output Files
- Notebook: /net/scratch2/smallyan/arithmetic_eval/no_exe_evaluation/code_critic_evaluation.ipynb
- JSON Summary: /net/scratch2/smallyan/arithmetic_eval/no_exe_evaluation/code_critic_summary.json
